File: ael_screen_ai/lib/services/auth_service.dart
import 'dart:convert';
import 'package:dio/dio.dart';
import '../config/app_config.dart';
import '../models/user_model.dart';
import 'api_client.dart';

class AuthService {
  final ApiClient _api = ApiClient();

  Future<Map<String, dynamic>> login(String email, String password) async {
    final resp = await _api.post('/auth/login', data: {
      'email': email,
      'password': password,
    });
    final data = resp.data['data'] as Map<String, dynamic>;
    await _api.setToken(data['access_token'] as String);
    return data;
  }

  Future<Map<String, dynamic>> register(
      String email, String password, String? displayName) async {
    final resp = await _api.post('/auth/register', data: {
      'email': email,
      'password': password,
      'display_name': displayName ?? email.split('@')[0],
    });
    final data = resp.data['data'] as Map<String, dynamic>;
    await _api.setToken(data['access_token'] as String);
    return data;
  }

  Future<Map<String, dynamic>> signInWithApple({
    required String identityToken,
    required String authorizationCode,
    required String userId,
    String? email,
    String? displayName,
  }) async {
    final resp = await _api.post('/auth/apple', data: {
      'identity_token': identityToken,
      'authorization_code': authorizationCode,
      'user_id': userId,
      'email': email,
      'display_name': displayName,
    });
    final data = resp.data['data'] as Map<String, dynamic>;
    await _api.setToken(data['access_token'] as String);
    return data;
  }

  Future<UserModel?> getProfile() async {
    try {
      final resp = await _api.get('/auth/me');
      final data = resp.data['data'] as Map<String, dynamic>;
      return UserModel.fromJson(data);
    } on DioException {
      return null;
    }
  }

  Future<void> logout() async {
    try {
      await _api.post('/auth/logout');
    } catch (_) {}
    await _api.clearToken();
  }

  Future<bool> isLoggedIn() async {
    final token = await _api.getToken();
    return token != null && token.isNotEmpty;
  }
}

File: backend/app/services/auth_service.py
"""Authentication service: Supabase auth + Apple Sign-In + JWT."""

from typing import Optional
from app.database import get_supabase, get_service_client
from app.config import settings
from app.models.user import UserResponse, AuthTokenResponse


async def register_user(email: str, password: str, display_name: Optional[str] = None) -> AuthTokenResponse:
    """Register a new user via Supabase Auth."""
    client = get_supabase()
    resp = client.auth.sign_up({
        "email": email,
        "password": password,
        "options": {"data": {"display_name": display_name or email.split("@")[0]}},
    })
    user = resp.user
    session = resp.session
    return AuthTokenResponse(
        access_token=session.access_token if session else "",
        refresh_token=session.refresh_token if session else None,
        user=UserResponse(
            id=user.id,
            email=user.email,
            display_name=user.user_metadata.get("display_name") if user.user_metadata else None,
        ),
    )


async def login_user(email: str, password: str) -> AuthTokenResponse:
    """Login with email and password."""
    client = get_supabase()
    resp = client.auth.sign_in_with_password({"email": email, "password": password})
    user = resp.user
    session = resp.session
    return AuthTokenResponse(
        access_token=session.access_token if session else "",
        refresh_token=session.refresh_token if session else None,
        user=UserResponse(
            id=user.id,
            email=user.email,
            display_name=user.user_metadata.get("display_name") if user.user_metadata else None,
        ),
    )


async def apple_sign_in(
    identity_token: str,
    authorization_code: str,
    user_id: str,
    email: Optional[str] = None,
    display_name: Optional[str] = None,
) -> AuthTokenResponse:
    """Sign in with Apple via Supabase Auth."""
    client = get_supabase()
    resp = client.auth.sign_in_with_id_token({
        "provider": "apple",
        "id_token": identity_token,
        "nonce": authorization_code,
    })
    user = resp.user
    session = resp.session
    # Update display name if provided
    if display_name and user:
        get_service_client().auth.admin.update_user_by_id(
            user.id, {"user_metadata": {"display_name": display_name}}
        )
    return AuthTokenResponse(
        access_token=session.access_token if session else "",
        refresh_token=session.refresh_token if session else None,
        user=UserResponse(
            id=user.id if user else user_id,
            email=email or (user.email if user else None),
            display_name=display_name,
        ),
    )


async def get_current_user(access_token: str) -> Optional[UserResponse]:
    """Get current user from access token."""
    try:
        client = get_supabase()
        user = client.auth.get_user(access_token)
        if not user or not user.user:
            return None
        u = user.user
        sub_info = get_service_client().table("subscriptions").select("*").eq("user_id", u.id).eq("status", "active").maybe_single().execute()
        is_premium = bool(sub_info.data) if sub_info else False
        expires = None
        if sub_info and sub_info.data:
            expires = sub_info.data.get("end_date")
        return UserResponse(
            id=u.id,
            email=u.email,
            display_name=u.user_metadata.get("display_name") if u.user_metadata else None,
            avatar_url=u.user_metadata.get("avatar_url") if u.user_metadata else None,
            created_at=u.created_at,
            is_premium=is_premium,
            subscription_expires=expires,
        )
    except Exception:
        return None


async def logout_user(access_token: str) -> bool:
    """Logout user."""
    try:
        client = get_supabase()
        client.auth.sign_out()
        return True
    except Exception:
        return False
